src/backend/distributed/executor/multi_server_executor.c

```diff
@@ -109,61 +106,80 @@ JobExecutorType(MultiPlan *multiPlan)
 									 "configured max_tracked_tasks_per_node limit")));
 		}
 	}
-	else if (executorType == MULTI_EXECUTOR_ROUTER)
+
+	return executorType;
+}
+
+
+/*
+ * RouterExecutablePlan returns whether a multi-plan can be executed using the
+ * router executor. Modify queries are always router executable, select queries
+ * are router executable only if executorType is real time.
+ */
+static bool
+RouterExecutablePlan(MultiPlan *multiPlan, MultiExecutorType executorType)
+{
+	Job *job = multiPlan->workerJob;
+	TaskType taskType = TASK_TYPE_INVALID_FIRST;
+	Query *masterQuery = multiPlan->masterQuery;
+	List *workerTaskList = job->taskList;
+	int taskCount = list_length(workerTaskList);
+	int dependedJobCount = list_length(job->dependedJobList);
+	Task *workerTask = NULL;
+	List *workerDependentTaskList = NIL;
+	bool masterQueryHasAggregates = false;
+
+	/* router executor cannot execute queries that hit more than one shard */
+	if (taskCount != 1)
 	{
-		Task *workerTask = NULL;
-		List *workerDependentTaskList = NIL;
-		bool masterQueryHasAggregates = false;
+		return false;
+	}
 
-		/* if we have repartition jobs with router executor, error out */
-		if (dependedJobCount > 0)
-		{
-			ereport(ERROR, (errmsg("cannot use router executor with repartition jobs"),
-							errhint("Set citus.task_executor_type to "
-									"\"task-tracker\".")));
-		}
+	/* check if the first task is a modify or a router task, short-circuit if so */
```

Comment:
If it's a short-circuit, why haven't we moved this check to the top? I think taskCount check could be below, what do you think?